### ISSUE TITLE
Using custom functions without sending a selection

### ISSUE THREAD

**Issue by SaskiaFreytag** [iSEE/iSEE#308]
Hi, 

Is there a way to make a custom function that does not use an interactive selection, but instead uses the row names or column names to select from?

**Comment by kevinrue** [comment]
Thanks for the question.

If I understand correctly, here is an example of what you're looking for:
a function that directly receives the row/column names from a user input rather than selections from other panels.

Note how the 2nd and 3rd arguments (`rows` and `columns`) are ignored. Those usually receive the interactive selections.
Instead, I've added additional arguments (`rownames` and `colnames`), that receive their value from the UI text box.
The value of those arguments are passed as a single character value which needs to be split.
See the [Deploying custom panels](https://bioconductor.org/packages/devel/bioc/html/iSEE.html) vignette for more information.
In this case, individual row/column names are separated by space character. See further below for an example of text input.
```
# For demonstration purposes, this function:
# - plots the TSNE results
# - highlights the selected column data points (= cells) in red
# - colors the other data points by the number of rows (= genes) with non-zero counts
CUSTOM_SELECTIONS <- function(
    se, rows, columns,
    rownames=character(0), colnames=character(0)
) {
    if (length(colnames) > 0) {
        colnames = strsplit(colnames, " ")[[1]]
    }
    if (length(rownames) > 0) {
        rownames = strsplit(rownames, " ")[[1]]
    }
    plot_data <- as.data.frame(reducedDim(se, "TSNE")[, 1:2, drop=FALSE])
    colnames(plot_data) <- paste0("TSNE", seq(1, 2))
    plot_data$colnames <- colnames(se) %in% colnames
    plot_data$rownames <- colSums(assay(sce, "counts")[rownames, , drop=FALSE] > 0)
    ggplot() +
        geom_point(aes(TSNE1, TSNE2, color=rownames), subset(plot_data, !colnames)) +
        geom_point(aes(TSNE1, TSNE2), subset(plot_data, colnames), color="red")
}
```
You can try it with our typical example dataset:
```
library(iSEE)
library(scRNAseq)
# Example data ----
sce <- ReprocessedAllenData()
class(sce)

library(scater)
counts(sce) <- assay(sce, "tophat_counts")
sce <- normalize(sce)

sce <- runPCA(sce, ncomponents=4)
sce <- runTSNE(sce)
rowData(sce)$ave_count <- rowMeans(counts(sce))
rowData(sce)$n_cells <- rowSums(counts(sce)>0)
sce

initialPanels <- DataFrame(
    Name=c("Custom data plot 1"),
    Width=c(6L))

customDataArgs <- customDataPlotDefaults(sce, 1)
customDataArgs$Function <- "CUSTOM_SELECTIONS"
customDataArgs$DataBoxOpen <- TRUE

app <- iSEE(
    se = sce,
    customDataArgs=customDataArgs,
    initialPanels=initialPanels,
    customDataFun=list(
        CUSTOM_SELECTIONS=CUSTOM_SELECTIONS
    ))
runApp(app)
```
An example of text input is:
```
colnames SRR2140059 SRR2139836 SRR2139694 SRR2139701 SRR2139318 SRR2140068
rownames Gm5414 Slc17a7 Mill1 Exoc1 Bloc1s6 Ttc39d
```
You can also test the function interactively with the following code:
```
CUSTOM_SELECTIONS(
    sce,
    colnames = paste(sample(colnames(se), 6), collapse = " "),
    rownames = paste(sample(rownames(se), 6), collapse = " ")
)
```
I hope this helps.

FYI, I haven't tested how well this scales with increasing numbers of columns/rows in the input text box. Please let us know!

**Comment by SaskiaFreytag** [comment]
This is great, but I was more hoping for a way that I could pass the entirety of the feature names or column names (probably possible for the column names with the current solution).

**Comment by kevinrue** [comment]
Well, my previous post was mostly a proof of concept that a custom panel can function without receiving selections from other panels, based on my understanding of your first post
> a custom function that does not use an interactive selection, but instead uses the row names or column names to select from

Then, depending how you implement the custom function, and what additional arguments you give it, that custom function can work on all the feature/columns, or whatever you want, really.
The key is to provide custom arguments (anything beyond the 3rd argument of the custom function) that are handled appropriately within the custom function. See https://github.com/kevinrue/iSEE_custom for examples/inspiration.

E.g. you could handle a special value `"all"` that tells the function to use all column/row names instead of an explicit subset of column/row names.
Obviously, you'd have to make sure that special value isn't a column/row name itself.

For the custom function above, that could be:
```
if (identical(colnames, "all")) {
    colnames <- colnames(se)
}
if (identical(rownames, "all")) {
    rownames <- rownames(se)
}
```

To be honest, I'm still not entirely sure of the exact behaviour you're aiming for, from the two descriptions that you gave above.
> a custom function that does not use an interactive selection, but instead uses the row names or column names to select from
> I was more hoping for a way that I could pass the entirety of the feature names or column names

If you can describe your use case again in further details, we may be able to give you more accurate advice/code.
Custom panel are a bit tricky to get familiar with, but they are extremely flexible and they should allow you to achieve your objective one way or another.

**Comment by SaskiaFreytag** [comment]
Hi Kevin,

I have implemented your solution. 

![Screen Shot 2019-07-30 at 9 53 35 am](https://user-images.githubusercontent.com/9170680/62094491-f327bf00-b2af-11e9-94ba-17ae9a860761.png)

However what I was really after was a dropdown menu, where I could select just one gene from a list. I am sorry I did not explain this better before. I see how what I wrote was super confusing.

Anyway here is my code. Any pointer would be super appreciated.

```
library(schex)
library(Seurat)
library(iSEE)

pbmc_small <- as.SingleCellExperiment(pbmc_small)
pbmc_small <- make_hexbin(pbmc_small, nbins=10, dimension_reduction = "PCA")

plot_hexbin_gene_new <- function(sce, rows=NULL, rownames=character(0), 
                                 columns=NULL, type="logcounts", action="prop_0"){

  plot_hexbin_gene(sce, type=type, gene=rownames, action=action)
}

schex_plot_gene <- customDataPlotDefaults(pbmc_small, 1)
schex_plot_gene$Function <- "plot_hexbin_gene_new"
schex_plot_gene$Arguments <- "type counts\naction prop_0\nrownames ODC1"
schex_plot_gene$ColumnSource <- "NULL"
schex_plot_gene$RowSource <- "NULL"
schex_plot_gene$DataBoxOpen <- TRUE

  
app <- iSEE(
    pbmc_small, 
    customDataArgs=schex_plot_gene, 
    initialPanels=DataFrame(
        Name=c("Custom data plot 1"),
        Width=c(12)),
    customDataFun=list(plot_hexbin_gene_new=plot_hexbin_gene_new)
    )


```

**Comment by kevinrue** [comment]
Indeed there is currently no way of adding UI inputs to custom panels, and I cannot foresee a solution to that: adding new UI inputs requires the addition of new observers, which basically equates to writing a new Shiny app. I have given a lot of thought about this during the development of custom panel and couldn't come up with any solution yet.

One possibility would be to use a "Column/Row statistics table" in the same way as a dropdown menu (i.e., selecting a row in the table and using the table filters).
That said, I just checked and tables cannot be sources of selections. I cannot immediately remember why that is.

Assuming this can be fixed, using a "Column statistics table 1" as a source of selection for a custom panel would look like the following.

```r
library(schex)
library(Seurat)
library(iSEE)

pbmc_small <- as.SingleCellExperiment(pbmc_small)
pbmc_small <- make_hexbin(pbmc_small, nbins=10, dimension_reduction = "PCA")

plot_hexbin_gene_new <- function(sce, rows=NULL, rownames=character(0), 
                                 columns=NULL, type="logcounts", action="prop_0"){
    
    plot_hexbin_gene(sce, type=type, gene=rownames, action=action)
}

schex_plot_gene <- customDataPlotDefaults(pbmc_small, 1)
schex_plot_gene$Function <- "plot_hexbin_gene_new"
schex_plot_gene$Arguments <- "type counts\naction prop_0\nrownames ODC1"
schex_plot_gene$ColumnSource <- "Column statistics table 1" # feature request: receive selection from table
schex_plot_gene$RowSource <- "---"
schex_plot_gene$DataBoxOpen <- TRUE

app <- iSEE(
    pbmc_small, 
    customDataArgs=schex_plot_gene, 
    initialPanels=DataFrame(
        Name=c("Custom data plot 1", "Column statistics table 1"),  # show table
        Width=c(8, 4)), # show the table beside the custom panel for convenience
    customDataFun=list(plot_hexbin_gene_new=plot_hexbin_gene_new)
)
```

**Comment by LTLA** [comment]
> adding new UI inputs requires the addition of new observers, which basically equates to writing a new Shiny app.

Shiny modules may or may not help, but in either case, it's all or nothing. Hypothetically, once you start writing a Shiny component, you're "all in", and you are expected to comply with `iSEE`'s internal observer infrastructure. And there are a lot of rules to follow.

> That said, I just checked and tables cannot be sources of selections. I cannot immediately remember why that is.

Well, you're only allowed to select one row at a time, so it wasn't all that useful to use as a _selection_ transmission. (Keeping in mind we have a different channel to transmit aesthetic parameters.) While DT allows selection multiple rows, that interferes with the user experience for other things, e.g., selecting genes to colour by becomes more painful if you have to unselect the previous gene every time.

I don't know of any case where I would be able to select multiple rows from a table without external knowledge of how they are important, e.g., from the metadata or from other panels. And if such information is present, that can be used directly for transmission.

Anyway, a better solution to @SaskiaFreytag's problem would be to support arbitrary observers. This is not impossible but requires some thought.

**Comment by SaskiaFreytag** [comment]
Thanks so much you guys for all your thoughts on this issue!

Even if I am not hundred percent happy with the current iSEE solution, I see that it would be super hard to allow custom UIs. Anyway I think iSEE is great!

**Comment by LTLA** [comment]
Should be closed by #327. This will allow you to make fully custom UIs. I believe @kevinrue uses **schex** as an example of creating a custom panel in his book, https://isee.github.io/iSEE-book/developing.html.
